components/domains/steps/checkoutCard.tsx

```diff
@@ -664,12 +669,76 @@ const CheckoutCard: FunctionComponent<CheckoutCardProps> = ({
   const onCurrencySwitch = (type: CurrencyType) => {
     if (type !== CurrencyType.ETH) setLoadingPrice(true);
     setDisplayedCurrency(type);
+    setReducedDuration(0);
+    setReducedDurationToken(null);
+    setBetterReducedDuration(0);
   };
 
   const onUpsellChoice = (enable: boolean) => {
     updateFormState({ isUpselled: enable });
   };
 
+  useEffect(() => {
+    const duration = formState.duration;
+    if (!invalidBalance || !priceInEth) {
+      setReducedDuration(0);
+      setReducedDurationToken(null);
+      setBetterReducedDuration(0);
+      return;
+    }
+    let solutionFoundWithCurrentToken = 0;
+    for (let newDuration = duration - 1; newDuration > 0; newDuration--) {
+      const newPriceInEth = getPriceForDuration(priceInEth, newDuration);
+      let newPrice = newPriceInEth;
+      if (displayedCurrency !== CurrencyType.ETH && quoteData)
+        newPrice = getDomainPriceAltcoin(quoteData.quote, newPriceInEth);
+      const balance = tokenBalances[displayedCurrency];
+      if (!balance) continue;
+      if (BigInt(balance) >= BigInt(newPrice)) {
+        setReducedDuration(newDuration);
+        solutionFoundWithCurrentToken = newDuration;
+        break;
+      }
+    }
+    // If we reach this point, the user doesn't have enough balance for any duration
+    // Try again but with other tokens than the displayed one
+    const tokens = Object.keys(tokenBalances).filter(
+      (token) => token !== displayedCurrency
+    );
+    (async () => {
+      for (const token of tokens) {
+        for (
+          let newDuration = duration;
+          newDuration > solutionFoundWithCurrentToken;
+          newDuration--
+        ) {
+          const newPriceInEth = getPriceForDuration(priceInEth, newDuration);
+          let newPrice = newPriceInEth;
+          if (token !== CurrencyType.ETH) {
+            const quoteData = await getTokenQuote(ERC20Contract[token]);
+            newPrice = getDomainPriceAltcoin(quoteData.quote, newPriceInEth);
+          }
+          const balance = tokenBalances[token];
+          if (!balance) continue;
+          if (BigInt(balance) >= BigInt(newPrice)) {
+            setBetterReducedDuration(newDuration);
+            setReducedDurationToken(token as CurrencyType);
```

Comment:
If we have a case where STRK token is better than ETH, but then let's say we have a 3rd token USDC, and you find another case, then you should ensure that `newDuration` is higher than `BetterReducedDuration` before adding it here no ?

Reply:
Ah yes I see, in fact I should simply swap both for loops, so that it will check for example
 - All tokens for 24 years
 - Then all for 23 years
 - Again, but find out that USDC allows 22 years and stop

Reply:
Yes perfect